idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+	if err != nil {
+		bld.WriteString("invalid")
+		return bld.String()
+	}
+	bld.WriteString(name)
+	return bld.String()
+}
+
+// setMetricName interns the MetricName in an
+// object store and stores the addresses of those strings
+// in MetricName.nodes
+func (mn *MetricName) setMetricName(name string) error {
+	nodes := strings.Split(name, ".")
+	mn.nodes = make([]uintptr, len(nodes))
+	for i, node := range nodes {
+		// TODO: add error checking? Fail somehow
+		nodePtr, err := IdxIntern.AddOrGet([]byte(node), false)
+		if err != nil {
+			log.Error("idx: Failed to acquire interned string for node name: ", err)
+			internError.Inc()
+			return fmt.Errorf("idx: Failed to acquire interned string for node %v, %v", node, err)
+		}
+		mn.nodes[i] = nodePtr
+	}
+
+	return nil
+}
+
+// ExtensionType is required to use custom marshaling as an extension
+// with msgp
+func (mn *MetricName) ExtensionType() int8 {
+	return 95
+}
+
+// Len is required to use custom marshaling as an extension
+// with msgp
+func (mn *MetricName) Len() int {
+	return len(mn.String())
+}
+
+// MarshalBinaryTo is required to use custom marshaling as an extension
+// in msgp
+func (mn *MetricName) MarshalBinaryTo(b []byte) error {
+	copy(b, []byte(mn.String()))
+	return nil
+}
+
+// UnmarshalBinary is required to use custom marshaling as an extension
+// in msgp
+func (mn *MetricName) UnmarshalBinary(b []byte) error {
+	err := mn.setMetricName(string(b))
+	return err
+}
+
+//msgp:shim TagKeyValue as:string using:TagKeyValue.createTagKeyValue/parseTagKeyValue
+
+// TagKeyValue stores a Key/Value pair. The strings
+// are interned in an object store before they are assigned.
+type TagKeyValue struct {
+	Key   uintptr
+	Value uintptr
+}
+
+// String returns a Key/Value pair in the form of
+// 'key=value'
+func (t *TagKeyValue) String() string {
+	var bld strings.Builder
+	return t.string(&bld)
+}
+
+func (t *TagKeyValue) string(bld *strings.Builder) string {
+	key, err := IdxIntern.GetStringFromPtr(t.Key)
+	if err != nil {
+		panic(fmt.Sprintf("idx: Failed to retrieve interned tag key: %v", err))
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		panic(fmt.Sprintf("idx: Failed to retrieve interned tag value: %v", err))
+	}
+
+	bld.WriteString(key)
+	bld.WriteString("=")
+	bld.WriteString(val)
+
+	return bld.String()
+}
+
+func (t TagKeyValue) createTagKeyValue() string {
+	bld := strings.Builder{}
+
+	key, err := IdxIntern.GetStringFromPtr(t.Key)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag key: ", err)
+		internError.Inc()
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag value: ", err)
+		internError.Inc()
+	}
+
+	bld.WriteString(key)
+	bld.WriteString("=")
+	bld.WriteString(val)
+
+	return bld.String()
+}
+
+func parseTagKeyValue(tag string) TagKeyValue {
+	var tkv TagKeyValue
+	if strings.Contains(tag, ";") {
```

Comment:
FYI there are more characters which are invalid: https://github.com/graphite-project/graphite-web/blob/7149908675664f341c91dcacfff83c55505564fa/docs/tags.rst#carbon
With this PR we'll get a validator for tag keys and values: https://github.com/grafana/metrictank/pull/1329/files#diff-991d213ed3916a1701643534978d4785R62